include/bit_algorithm_details.hpp

```diff
@@ -1,29 +1,96 @@
+// License: BSD 3-Clause License
 // ========================================================================== //
 #ifndef _BIT_ALGORITHM_DETAILS_HPP_INCLUDED
 #define _BIT_ALGORITHM_DETAILS_HPP_INCLUDED
 // ========================================================================== //
 
 
-
-// ================================ PREAMBLE ================================ //
+// ============================== PREAMBLE ================================== //
 // C++ standard library
 // Project sources
-#include "bit.hpp"
 // Third-party libraries
 // Miscellaneous
-namespace bit {
-// ========================================================================== //
 
+namespace bit {
+/*
+ * Takes two words and merges them together.
+ * 
+ * @param w1 the first operand of the merge
+ * @param w2 the second operand of the merge
+ * @param w1_position the position within w1 to start the merge from
+ * @return a word given by merging w1 and w2
+ */ 
+template <class T>
+T _shift_and_merge(T w1, T w2, std::size_t w1_position) {
+
+  if (w1_position == 0) {
+    return w1;
+  }
+
+  constexpr std::size_t num_digits = bit::binary_digits<T>::value;
+
+  T tmp1 = w1 << w1_position;
+  T tmp2 = w2 >> (num_digits - w1_position);
+  return tmp1 | tmp2;
+}
 
+/*
+ * This function reads a T's worth of bits starting from wherever
+ * the start bit_iterator is pointing. This means that we might be making
+ * a word read from words which span multiple words.
+ *
+ * @param start the bit_iterator to start reading from
+ * @param guard the furthest position to read to, ensuring there are not
+ * memory access violations
+ * @return a word read starting from the start position
+ */
+template <class WrappedIter, class T = typename WrappedIter::word_type>
+T _read_word_raw(WrappedIter start, WrappedIter guard) {
+  const unsigned long dist = std::distance(start, guard);
+
+  if (dist < bit::binary_digits<T>::value) {
+    /* we don't have sufficient bits to read so 
+       instead of accessing bad memory, just pad 
+       the remainder of the read with 0s */
+    return *(start.base()) << start.position();
+  } else {
+    // we have enough bits to read to form a word
+    T lhs = *(start.base());
+    T rhs = *(start.base() + 1);
+    return _shift_and_merge(lhs, rhs, start.position());
+  }
+} 
+
+/*
+ * Allows the user to set or unset bits in range [begin, end) of the word at base
+ *
+ * @param begin the bit from which we start setting/unsetting 
+ * @param end the bit from which we stop setting bits
+ * @param base an iterator to the bit_value we wish to change
+ * @param bv the bit value which we wish to set all bits in the range to.
+ */
+template <class BitIter>
+void _set_within_word(BitIter first, BitIter last, 
+    typename BitIter::iterator_type base, bit::bit_value bv) {
+    using word_type = typename BitIter::word_type;
+    
+    word_type mask = (static_cast<word_type>(1) << (last - first)) - 1;
+    mask <<= first.position();
+
+    if (bv == bit1) {
```

Comment:
This is most likely an extremely minor thing, but is it worth thinking about avoiding the `if/else` condition and replacing it with a single statement? I don't know much about branching, but I figure if we can avoid if/else why not.